### ISSUE TITLE
Certain images are ignored when creating AtlasTextures with Import Mode set to 'Mesh2D' 

### ISSUE THREAD

**Issue by LakshayaG73** [godotengine/godot#68350]
### Godot version

v4.0.beta4.official [e6751549c]

### System information

Windows 11

### Issue description

If we create an AtlasTexture from sprites, and set the 'Import Mode' to Mesh2D, certain sprites will be ignored, and these sprites will show up as empty frames if you try to use them in an AnimatedSprite2D. 

I've tested this, and it seems only certain images are being ignored, though I'm not sure why. For certain animations, images are ignored, whereas for others 0 are ignored. The same images are ignored every time, making me think that it is some property of the images themselves. 

### Steps to reproduce

Import sprites into project. 

Create AtlasTexture with import mode 'Mesh2D'

Some images will be missing if you try to use them in an AnimatedSprite2D. 

### Minimal reproduction project

[MRP_AtlasTexture.zip](https://github.com/godotengine/godot/files/9947164/MRP_AtlasTexture.zip)

**Comment by LakshayaG73** [comment]
[MRP_AtlasTexture.zip](https://github.com/godotengine/godot/files/9947164/MRP_AtlasTexture.zip)

**Comment by kleonc** [comment]
> ### Godot version
> 
> 4.0 latest

Please be more specific, "4.0 latest" will mean something else in the future. You can copy the exact version from the editor:
![Godot_v4 0-beta4_win64_SuC9MjZoHY](https://user-images.githubusercontent.com/9283098/200205544-2cd0c7c6-ff12-499d-8041-620223d067a7.png) gives: v4.0.beta4.official [e6751549c]



> ### Minimal reproduction project
> 
> Uploading MRP_AtlasTexture.zip…

You could have just edited it instead of adding another comment.


I've already edited the description of yours, if that's not the version you meant please re-edit.

---

Regarding the issue: can confirm, there's clearly an issue when importing atlas in 'Mesh2D' mode. Will investigate further.

**Comment by AThousandShips** [comment]
It looks like there's a problem with negative vertices somewhere, causing the packing to break, will take a deeper look

**Comment by AThousandShips** [comment]
Might have found the issue, will explore some more, it seems the generation of polygons from the opaque texture bugs in some contexts, seemingly lowering the delta improves the results, but will see if it's reliable, it does however solve the problem for this specific case

**Comment by AThousandShips** [comment]
Comparing with the code for Sprite2D and generating meshes I think minor changes to the settings can help, and possibly adding settings in the importer for grow/shrink/simplification to match

**Comment by AThousandShips** [comment]
I will take a look at adding settings to the importer and using the same defaults as Sprite2D which in my testing of this bug solves it

There is still the problem with the texture gaps which I haven't been able to identify yet:
![three](https://user-images.githubusercontent.com/96648715/200320864-22aaeef3-795f-4e06-a5e4-515ac0699d88.png)

**Comment by AThousandShips** [comment]
Opened a PR for this, note however that this does not include any fix for the loading so that bug will remain if you do not add a PR that solves that

**Comment by LakshayaG73** [comment]
> Opened a PR for this, note however that this does not include any fix for the loading so that bug will remain if you do not add a PR that solves that

Thank you.

I will apply both your PRs then. This solves the issue completely?

**Comment by AThousandShips** [comment]
It does not solve the gap problem, will investigate it further later, currently I am going to have a rest!

Been lovely engaging with these bugs and always lovely to meet other Godot users

**Comment by LakshayaG73** [comment]
Well deserved and you have unblocked my entire team, so I cannot convey enough thanks.

What is the gap problem? Does it distort the textures? Should I still avoid using the mesh mode?

**Comment by AThousandShips** [comment]
It is where the pixels are missing as in the image above, so in these cases it is currently not generating correct results, so avoid for now, I hope it won't make the resulting images too big for you when missing the gains in packing, hopefully that bug can be identified and solved too before too long

**Comment by LakshayaG73** [comment]
> It is where the pixels are missing as in the image above, so in these cases it is currently not generating correct results, so avoid for now, I hope it won't make the resulting images too big for you when missing the gains in packing, hopefully that bug can be identified and solved too before too long

The gains from Mesh packing cannot be ignored. It will greatly help my game and several others, i'm sure. 

Shall I create a new issue for this?

**Comment by AThousandShips** [comment]
Go ahead!

**Comment by AThousandShips** [comment]
It appears the persistent problem now isn't with the polygons, after rendering out the mesh after realising I could visualise it, and I now suspect it lies with the plotting out of the triangles, will test some alternative methods for drawing it out later today and hopefully it solves both issues

**Comment by AThousandShips** [comment]
It turns out the clipping problem was far easier to solve than I had thought, a simple check that wasn't working right

I am still working on the gaps but I will update the PR for the base fix and continue working on identifying the gaps problem, while this won't be helpful for production the prototyping with those gaps at least for prototyping purposes it should be acceptable I hope

**Comment by AThousandShips** [comment]
@LakshayaG73 packing should now be properly fixed with no deformed or clipped meshes, can you confirm? (For example no cut legs on some of the characters) and I will continue to investigate the Tale of the Lost Pixels

**Comment by LakshayaG73** [comment]
can confirm. I don't see any issues with the packing (although I never did. images were either skipped completely or not at all). 

when you say 'gaps', are you referring to these two lines? 

![image](https://user-images.githubusercontent.com/1243371/200691426-5f41e49b-b466-4920-aa00-495c38ebf36c.png)

**Comment by AThousandShips** [comment]
In your MRP if you look at the bottom the legs of the characters in One are clipped

Those are the gaps yes

**Comment by AThousandShips** [comment]
I have identified some underlying causes of the original polygon issues, but so far nothing in the filling area, will prioritize the polygons first

**Comment by LakshayaG73** [comment]
Hi @AThousandShips I can still repro this in 4.2. Importing AtlasTextures as Mesh2D is completely broken. The AtlasTexture seems to be created, but using any of the sprites from the AtlasTexture is completely broken. The images don't show up.

**Comment by AThousandShips** [comment]
This is a known different bug, not this bug, see:
* https://github.com/godotengine/godot/issues/85892

**Comment by LakshayaG73** [comment]
That issue is related to the files not loading cleanly post import. But post restart, the AtlasTexture still functions. I have opened that issue.

This bug may be fixed, but 'Mesh2D' as an AtlasTexture still does not work. Shall I open a new issue for the same?

**Comment by AThousandShips** [comment]
Indeed, my bad missed it was your report, *this* bug has been fixed and the area changed hasn't been reverted, so it's a new bug with similar symptoms
